clientapi/routing/sso.go

```diff
@@ -0,0 +1,203 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"encoding/xml"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/matrix-org/dendrite/clientapi/auth"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/internal/config"
+	"github.com/matrix-org/dendrite/internal/sqlutil"
+	"github.com/matrix-org/dendrite/userapi/storage/accounts"
+	"github.com/matrix-org/util"
+)
+
+// the XML response structure of CAS ticket validation
+type casValidateResponse struct {
+	XMLName               xml.Name `xml:"serviceResponse"`
+	Cas                   string   `xml:"cas,attr"`
+	AuthenticationSuccess struct {
+		User string `xml:"user"`
+	} `xml:"authenticationSuccess"`
+}
+
+// SSORedirect implements GET /login/sso/redirect
+// https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-login-sso-redirect
+// If the incoming request doesn't contain a SSO token, it will redirect to the SSO server
+// Else it will validate the SSO token, and redirect to the "redirectURL" provided with an extra "loginToken" param
+func SSORedirect(
+	req *http.Request,
+	accountDB accounts.Database,
+	cfg *config.ClientAPI,
+) util.JSONResponse {
+	// If dendrite is not configured to use SSO by the admin return bad method
+	if !cfg.CAS.Enabled || cfg.CAS.Server == "" {
+		return util.JSONResponse{
+			Code: http.StatusNotImplemented,
+			JSON: jsonerror.NotFound("Method disabled"),
+		}
+	}
+
+	// A redirect URL is required for this endpoint
+	redirectURLStr := req.FormValue("redirectUrl")
+	if redirectURLStr == "" {
+		return util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.MissingArgument("redirectUrl parameter missing"),
+		}
+	}
+	// Check if the redirect url is a valid URL
+	redirectURL, err := url.Parse(redirectURLStr)
+	if err != nil {
+		return util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.InvalidArgumentValue("Invalid redirectURL: " + err.Error()),
+		}
+	}
+
+	// If the request has a ticket param, validate the ticket instead of redirecting to SSO server
+	if ticket := req.FormValue("ticket"); ticket != "" {
+		return ssoTicket(req, redirectURL, accountDB, cfg)
+	}
+
+	// Adding the params to the sso url
+	ssoURL := cfg.CAS.URL
+	ssoQueries := make(url.Values)
+	// the service url that we send to CAS is homeserver.com/_matrix/client/r0/login/sso/redirect?redirectUrl=xyz
+	ssoQueries.Set("service", req.RequestURI)
```

Comment:
I suspect it doesn't matter too much, but it could be clearer to redirect to a different endpoint instead of using `/login/sso/redirect` as both the redirect endpoint and the endpoint that resolves the ticket. I'm not sure if this is common in CAS implementations or not. Synapse uses a separate endpoint, but it is also under the `/_matrix` prefix, which it should *not* be.

Reply:
Yeah, I thought about this too and in the older version of the spec ([r0.4.0](https://matrix.org/docs/spec/client_server/r0.4.0#cas-based-client-login)) there are 2 endpoints specified for CAS auth (`/login/cas/redirect` and `/login/cas/ticket`). But the newer spec ([r0.6.1](https://matrix.org/docs/spec/client_server/r0.6.1#sso-client-login)) doesn't mention the separate `/ticket` endpoint, so I thought I should just go with the single endpoint. 

Splitting it up into 2 endpoints can be easily done. 

About it not being in the `/_matrix` prefix, I don't know enough about common practices in other CAS implementations to comment. The spec does say `/_matrix/client/r0/login/sso/redirect` though.

Reply:
> Yeah, I thought about this too and in the older version of the spec ([r0.4.0](https://matrix.org/docs/spec/client_server/r0.4.0#cas-based-client-login)) there are 2 endpoints specified for CAS auth (`/login/cas/redirect` and `/login/cas/ticket`). But the newer spec ([r0.6.1](https://matrix.org/docs/spec/client_server/r0.6.1#sso-client-login)) doesn't mention the separate `/ticket` endpoint, so I thought I should just go with the single endpoint.

I believe the second endpoint is meant to be implementation specific!

> Splitting it up into 2 endpoints can be easily done.

I'd let @Kegsay or @neilalexander decide whether this should be done or not. 😄 

> About it not being in the `/_matrix` prefix, I don't know enough about common practices in other CAS implementations to comment. The spec does say `/_matrix/client/r0/login/sso/redirect` though.

Sorry for not being clear, I meant the ticket endpoint. The redirect endpoint must be  `/_matrix/client/r0/login/sso/redirect` as you noted.

Reply:
> I'd let @Kegsay or @neilalexander decide whether this should be done or not

Yeah I agree :sweat_smile: 

> I meant the ticket endpoint

The ticket endpoint is `/_matrix/client/r0/login/cas/ticket` according to the [old spec](https://matrix.org/docs/spec/client_server/r0.4.0#get-matrix-client-r0-login-cas-ticket), that could explain why synapse has it under `/_matrix` as well.

Reply:
2 endpoints please, both under `/_matrix` will do. If we don't put it under `/_matrix` then we need to mess around with routers more, and the only reason why we don't want it there is "just in case" the spec suddenly uses that endpoint for something, which feels unlikely.